libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/internal/MapboxNavigationViewApi.kt

```diff
@@ -0,0 +1,105 @@
+package com.mapbox.navigation.dropin.internal
+
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.geojson.Point
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.dropin.NavigationViewApi
+import com.mapbox.navigation.ui.app.internal.Store
+import com.mapbox.navigation.ui.app.internal.destination.Destination
+import com.mapbox.navigation.ui.app.internal.destination.DestinationAction
+import com.mapbox.navigation.ui.app.internal.extension.dispatch
+import com.mapbox.navigation.ui.app.internal.navigation.NavigationState
+import com.mapbox.navigation.ui.app.internal.navigation.NavigationStateAction
+import com.mapbox.navigation.ui.app.internal.routefetch.RoutePreviewAction
+import com.mapbox.navigation.ui.app.internal.routefetch.RoutePreviewState
+import com.mapbox.navigation.ui.app.internal.routefetch.RoutesAction
+import com.mapbox.navigation.ui.app.internal.tripsession.TripSessionStarterAction
+import com.mapbox.navigation.ui.app.internal.endNavigation as endNavigationAction
+
+@ExperimentalPreviewMapboxNavigationAPI
+internal class MapboxNavigationViewApi(
+    private val store: Store
+) : NavigationViewApi() {
+
+    override fun fetchRoutes(options: RouteOptions) {
+        store.dispatch(RoutePreviewAction.FetchOptions(options))
+    }
+
+    override fun fetchRoutes(points: List<Point>) {
+        store.dispatch(RoutePreviewAction.FetchPoints(points))
+    }
+
+    override fun setPreviewRoutes(routes: List<NavigationRoute>) {
+        store.dispatch(RoutePreviewAction.Ready(routes))
+    }
+
+    override fun setRoutes(routes: List<NavigationRoute>) {
+        store.dispatch(RoutesAction.SetRoutes(routes))
+    }
+
+    override fun setDestination(point: Point) {
+        store.dispatch(DestinationAction.SetDestination(Destination(point)))
+    }
+
+    override fun enableTripSession() {
+        store.dispatch(TripSessionStarterAction.EnableTripSession)
+    }
+
+    override fun enableReplaySession() {
+        store.dispatch(TripSessionStarterAction.EnableReplayTripSession)
+    }
+
+    override fun isReplayEnabled(): Boolean {
+        return store.state.value.tripSession.isReplayEnabled
+    }
+
+    override fun startFreeDrive() {
+        store.dispatch(RoutesAction.SetRoutes(emptyList()))
+        store.dispatch(RoutePreviewAction.Ready(emptyList()))
+        store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
```

Comment:
We should also set the destination to null

Reply:
Let's change this function to simply transition to `FreeDrive`

```
override fun startFreeDrive() {
    store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
}
```

Move the other lines to `NavigationViewApi.startFreeDrive()` as an extension.

```
NavigationViewApi.startFreeDrive() {
    store.dispatch(DestinationAction.SetDestination(null))
    store.dispatch(RoutesAction.SetRoutes(emptyList()))
    store.dispatch(RoutePreviewAction.Ready(emptyList()))
    startFreeDrive()
}
```

Reply:
> Let's change this function to simply transition to FreeDrive
> ```
> override fun startFreeDrive() {
>     store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
> }
> ```

Isn't calling the proposed `startFreeDrive()` method to simply transition to the `FreeDrive` risk putting NavigationView in a corrupted state?
My understanding is that NavigationView should:
- enter **FreeDrive** when there is no `Destination` set and both `previewRoutes` and MapboxNavigation `routes` are empty
- enter **DestinationPreview** state when `Destination` is set and both `previewRoutes` and MapboxNavigation `routes` are empty
- enter **RoutePreview** state when `Destination` is set, `previewRoutes` is not empty and MapboxNavigation `routes` is empty
- enter **ActiveNavigation** state when `Destination` is set, `previewRoutes` and MapboxNavigation `routes` are not empty
- enter **Arrival** state when `Destination` is set, `previewRoutes` and MapboxNavigation `routes` are not empty, and `ArrivalObserver.onFinalDestinationArrival()` is called

<img width="931" alt="Screen Shot 2022-08-16 at 10 00 03 AM" src="https://user-images.githubusercontent.com/2678039/184899543-fd0f175e-6d8d-4774-9fe5-cf25e70cce99.png">

> Move the other lines to NavigationViewApi.startFreeDrive() as an extension.
> 
> ```
> NavigationViewApi.startFreeDrive() {
>     store.dispatch(DestinationAction.SetDestination(null))
>     store.dispatch(RoutesAction.SetRoutes(emptyList()))
>     store.dispatch(RoutePreviewAction.Ready(emptyList()))
>     startFreeDrive()
> }
> ```

You can't do that for a couple of reasons.
1. Proposed extension function doesn't have access to the `store`.
2. _If a class has a member function, and an extension function is defined which has the same receiver type, the same name, and is applicable to given arguments, the member always wins._ ([source](https://kotlinlang.org/docs/extensions.html#extensions-are-resolved-statically))

Reply:
The idea is to basically replicate what you are doing for `startRoutePReview` and `startActiveNavigation`.  

**Inside NavigationViewApi**
```
NavigationViewApi.startFreeDrive() {
    setDestination(null)
    setPreviewRoutes(empty)
    setRoutes(empty)
    startFreeDrive()
}
```

**Inside MapboxNavigationViewApi**
```
fun startFreeDrive() {
    // check destination is null
    // check route preview is null
    // check routes is null

    store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
}
```

Reply:
```kotlin
fun MapboxNavigationViewApi.startFreeDrive() {
    // check destination is null
    // check route preview is null
    // check routes is null

    store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
}
```

What's the benefit of having a method (^) that will always fail if the below methods are not called first?
```kotlin
setDestination(null)
setPreviewRoutes(empty)
setRoutes(empty)
```

Isn't it NavigationView responsibility to correctly update its internal state before moving to FreeDrive state?

```kotlin
fun MapboxNavigationViewApi.startFreeDrive() {
  setDestination(null)
  setPreviewRoutes(empty)
  setRoutes(empty)
  store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
}